### PR TITLE
[NPU]Support GLM5.2 and FP8 DSA&Indexer kvcache for 950

### PR DESCRIPTION
**Pull request by LinyuanLi0046**
## Motivation

Enable GLM-5.2 inference on 950PR/DT NPU with FP8 KV cache and MLAProlog, while reducing redundant indexer cache allocation for GLM5.2.

## Modification

- Add quant sparse attention and quantized DSA indexing for FP8 KV cache.
- Support packed FP8 DSA&indexer KV cache through the existing --kv-cache-dtype option
- Add mlaprolog support for GLM5.2 on 950NPU and adapt MLAProlog weight/scale preparation for BF16 and MXFP8 projections, and align RoPE layouts and KV cache writes across DSA execution paths.
- Introduce compact indexer layer-to-slot mapping with matching memory accounting.
- Account for independent target/draft KV cache dtypes and update MTP cache relocation and CPU offload to preserve packed KV data and indexer scales.

## Accuracy Tests

```
gpqa_diamond report table:
┌──────────────────────┬──────────────┬────────────┬──────────┬───────┬─────────┐
│ Model                │ Dataset      │ Metric     │ Subset   │   Num │ Score   │
├──────────────────────┼──────────────┼────────────┼──────────┼───────┼─────────┤
│ GLM-5.2-W8A8C8-mxfp8 │ GPQA-Diamond │ Accuracy ↑ │ default  │   198 │ 93.9%   │
└──────────────────────┴──────────────┴────────────┴──────────┴───────┴─────────┘ 

2026-09-08 07:17:28 - evalscope - INFO: Skipping report analysis (`analysis_report=False`).                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       
```

## Speed Tests and Profiling

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.




<!-- pr-states:start -->
---
### CI States

Latest PR Test (Base): <!-- slot:pr-test:start -->:white_check_mark: [Run #34343316959](https://github.com/sgl-project/sglang/actions/runs/34343316959)<!-- slot:pr-test:end -->
Latest PR Test (Extra): <!-- slot:pr-test-extra:start -->:x: [Run #34343316745](https://github.com/sgl-project/sglang/actions/runs/34343316745)<!-- slot:pr-test-extra:end -->
Latest PR Test (AMD ROCm 10): <!-- slot:pr-test-amd-rocm720:start -->:x: [Run #34343316954](https://github.com/sgl-project/sglang/actions/runs/34343316954)<!-- slot:pr-test-amd-rocm720:end -->
<!-- pr-states:end -->

### CODE REVIEW

File: python/sglang/srt/utils/common.py [hunk 1]
```diff
@@ -857,6 +857,14 @@ def get_device_name(device_id: int = 0) -> str:
         return torch.npu.get_device_name(device_id)
 
 
+def is_950_npu(device_id: int = 0) -> bool:
```
Comment (AndyLi429) [inline python/sglang/srt/utils/common.py]: use latest code and is_npu_arch35

File: python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py [hunk 2]
```diff
@@ -99,8 +100,25 @@ def __init__(
...
-            torch.float
+        q_b_scale = getattr(self.q_b_proj, "weight_scale", None)
+        self.q_b_proj_weight_scale = (
+            q_b_scale.view(1, -1).to(torch.float) if q_b_scale is not None else None
+        )
+        self.is_950 = is_950_npu(torch.npu.current_device())
+
+    def uses_mlaprolog(self) -> bool:
+        pool = get_token_to_kv_pool()
+        if self.is_950 and pool.index_head_dim is not None:
```
Comment (AndyLi429) [inline python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py]: all DSA model will use mlaprolog?

File: python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py [hunk 3]
```diff
@@ -447,17 +551,46 @@ def forward_mlaprolog(self, positions, hidden_states, forward_batch):
...
+        elif self.weight_quant_mode == 1:
+            mla_prolog_input_args["dequant_scale_w_uq_qr"] = self.q_b_proj_weight_scale
+        if packed:
+            mla_prolog_input_args.update(
+                ckvkr_repo_mode=1, quant_scale_repo_mode=1, tile_size=128
+            )
+        import torch_npu
+
+        if self.is_950 or hasattr(torch_npu, "npu_mla_prolog_v3"):
+            prolog = torch_npu.npu_mla_prolog_v3
```
Comment (AndyLi429) [inline python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py]: use native ops?no customs ops available for 950?

File: python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py (hunk 2 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py]: This function only is used when USE_MLAPO env var is true. If this env var is not set, this function is not useded

Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py]: since MLAPO op in kernel-npu repo is not availabe on 950

File: python/sglang/srt/utils/common.py (hunk 1 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/utils/common.py]: I will fix this

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py [hunk 4]
```diff
@@ -1143,28 +1143,58 @@ def forward_sparse(
...
-                ),
-                block_table=self.forward_metadata.block_tables,
-                sparse_block_size=1,
-                layout_query="TND",
-                layout_kv="PA_BSND",
-                sparse_mode=3,
-                attention_mode=2,
-                return_softmax_lse=False,
-            )
+            if self.token_to_kv_pool.dsa_kv_cache_store_fp8:
```
Comment (iforgetmyname) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py]: use --kvcache-dtype fp8

Comment (LinyuanLi0046) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py]: should controled by -kv-cache-dtype fp8

File: python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py (hunk 3 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py]: use torch npu's mlaprolog

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py (hunk 4 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py]: fixed

File: python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py [hunk 5]
```diff
@@ -241,13 +258,82 @@ def preprocess_weights(self, hidden_states):
...
+                qkv_scale[: self.q_lora_rank].contiguous().view(torch.float8_e8m0fnu)
+            )
+            self.qkv_a_proj_scale_kv = (
+                qkv_scale[self.q_lora_rank :].contiguous().view(torch.float8_e8m0fnu)
+            )
+            self.q_b_proj_scale = qb_scale.contiguous().view(torch.float8_e8m0fnu)
+            self.q_b_proj_weight = npu_format_cast(
+                self.q_b_proj.weight.data.contiguous()
+            )
+            self.weight_quant_mode = 3
```
Comment (iforgetmyname) [inline python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py:306]: what is this weight_quant_mode?
give this a clearer definition

File: python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py [hunk 6]
```diff
@@ -470,6 +598,8 @@ def forward_mlaprolog(self, positions, hidden_states, forward_batch):
         )
 
     def forward(self, positions, hidden_states, forward_batch, zero_allocator):
+        if self.uses_mlaprolog():
+            return self.forward_mlaprolog(positions, hidden_states, forward_batch)
```
Comment (iforgetmyname) [inline python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py]: this condition checking is too general
we do have a `forward_mlaprolog` branch under, merge them into one

File: python/sglang/srt/hardware_backend/npu/modules/deepseek_v2_attention_mla_npu.py [hunk 7]
```diff
@@ -357,7 +371,13 @@ def forward_dsa_prepare_npu(
     prev_topk_indices: torch.Tensor = None,
 ):
     dynamic_scale = None
-    if is_mla_preprocess_enabled() and forward_batch.forward_mode.is_decode():
+    if not hasattr(m, "_npu_is_arch35"):
+        m._npu_is_arch35 = is_npu_arch35()
```
Comment (iforgetmyname) [inline python/sglang/srt/hardware_backend/npu/modules/deepseek_v2_attention_mla_npu.py]: not used

File: python/sglang/srt/hardware_backend/npu/modules/deepseek_v2_attention_mla_npu.py [hunk 8]
```diff
@@ -505,33 +536,31 @@ def forward_dsa_core_npu(
...
     attn_output = attn_output.view(-1, m.num_local_heads, m.kv_lora_rank)
 
-    attn_bmm_output = torch.empty(
-        (attn_output.shape[0], m.num_local_heads, m.v_head_dim),
-        dtype=attn_output.dtype,
-        device=attn_output.device,
-    )
-
-    if (
+    if m._npu_is_arch35 or (
```
Comment (iforgetmyname) [inline python/sglang/srt/hardware_backend/npu/modules/deepseek_v2_attention_mla_npu.py]: use _is_npu_arch35 = is_npu_arch35()

File: python/sglang/srt/hardware_backend/npu/quantization/linear_method_npu.py [hunk 9]
```diff
@@ -256,22 +256,33 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
     def apply(
         self,
         layer: torch.nn.Module,
-        x: torch.Tensor,
+        x: torch.Tensor | tuple[torch.Tensor, torch.Tensor],
```
Comment (iforgetmyname) [inline python/sglang/srt/hardware_backend/npu/quantization/linear_method_npu.py]: `from typing import Tuple`
`Tuple[torch.Tensor, torch.Tensor]`

File: python/sglang/srt/hardware_backend/npu/memory_pool_npu.py [hunk 10]
```diff
@@ -553,6 +555,42 @@ def __init__(
...
+        assert all(
+            self.start_layer <= i < self.start_layer + self.layer_num
+            for i in self.indexer_layer_ids
+        )
+        requested_kv_cache_dim = kv_cache_dim
+        self.dsa_kv_cache_store_fp8 = (
+            index_head_dim is not None
+            and dtype == torch.float8_e4m3fn
+            and requested_kv_cache_dim is not None
+        )
```
Comment (iforgetmyname) [inline python/sglang/srt/hardware_backend/npu/memory_pool_npu.py:585]: also respect --kvcache-dtype

File: python/sglang/srt/mem_cache/kv_cache_configurator.py [hunk 11]
```diff
@@ -1504,20 +1504,62 @@ def _build_ascend_mla_kv_pool(
...
+                layer_id
+                for layer_id in range(
+                    self.layer_info.start_layer,
+                    self.layer_info.end_layer,
+                )
+                if not dsa_layer_skips_topk(self.model_config.hf_config, layer_id)
+            )
+        use_dsa_fp8_kv_cache_storage = (
+            self.kv_cache_dtype == torch.float8_e4m3fn and is_arch35
+        )
```
Comment (iforgetmyname) [inline python/sglang/srt/mem_cache/kv_cache_configurator.py:1542]: remove these checks

File: python/sglang/srt/mem_cache/kv_cache_configurator.py [hunk 12]
```diff
@@ -1504,20 +1504,62 @@ def _build_ascend_mla_kv_pool(
...
+            logger.info(
+                "NPU DSA indexer layout: %s (%d/%d layers)",
+                (
+                    "compact"
+                    if token_to_kv_pool.num_indexer_layers < token_to_kv_pool.layer_num
+                    else "all-layer"
+                ),
+                token_to_kv_pool.num_indexer_layers,
+                token_to_kv_pool.layer_num,
+            )
```
Comment (iforgetmyname) [inline python/sglang/srt/mem_cache/kv_cache_configurator.py]: remove this log

File: python/sglang/srt/model_executor/pool_configurator.py
Comment (iforgetmyname) [inline python/sglang/srt/model_executor/pool_configurator.py:1]: review the necessity of changing this file?

**Review by iforgetmyname — CHANGES_REQUESTED** [review]

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py (hunk 4 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py]: fixed

File: python/sglang/srt/model_executor/pool_configurator.py
Comment (LinyuanLi0046) [reply python/sglang/srt/model_executor/pool_configurator.py:1]: I've revised the changes to pool_configurator.py and removed unnecessary modifications. The implementation now reuses the existing shared logic from the GPU path. It also no longer handles separate capacity accounting for the less common configuration where the target uses FP8 KV cache and the draft uses BF16 KV cache.
The remaining changes are limited to:
1. Correcting NPU indexer memory accounting. The existing GPU-oriented calculation assumes FP8 indexer storage, which underestimates memory usage for BF16 indexers. The NPU path needs to account for the actual storage dtype.
2. Using all-layer indexer capacity accounting on non-arch35 NPUs, matching their actual allocation layout.

File: python/sglang/srt/mem_cache/kv_cache_configurator.py (hunk 12 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/mem_cache/kv_cache_configurator.py]: removed

File: python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py (hunk 5 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py:306]: weight quant mode 3 infers weights(weight_dq, weight_uq_qr, weight_uk, weight_dkv_kr)' dtype is mxfp8

File: python/sglang/srt/hardware_backend/npu/modules/deepseek_v2_attention_mla_npu.py (hunk 7 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/modules/deepseek_v2_attention_mla_npu.py]: fixed

File: python/sglang/srt/hardware_backend/npu/modules/deepseek_v2_attention_mla_npu.py (hunk 8 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/modules/deepseek_v2_attention_mla_npu.py]: fixed

File: python/sglang/srt/hardware_backend/npu/quantization/linear_method_npu.py (hunk 9 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/quantization/linear_method_npu.py]: fixed

File: python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py (hunk 6 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py]: fixed

File: python/sglang/srt/hardware_backend/npu/memory_pool_npu.py (hunk 10 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/memory_pool_npu.py:585]: this may leads to error if set draft kvcache dtype=BF16 when target model‘s kvcche dtype is FP8E4M3

File: python/sglang/srt/mem_cache/kv_cache_configurator.py (hunk 11 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/mem_cache/kv_cache_configurator.py:1542]: removed glm arch check

**Review by iforgetmyname — APPROVED** [review]
